File: LeetCode/Python Solutions/Shortest Bridge/shortest.py
from typing import List
from collections import deque

class Solution:
    def shortestBridge(self, grid: List[List[int]]) -> int:

        # DFS function to traverse and mark the first island
        def dfs(i, j):
            # Ignore if the current cell is out of bounds or is water or has been visited
            if not(0 <= i < len(grid) and 0 <= j < len(grid[0]) and grid[i][j] not in (0, -1)):
                return
            # Mark the current cell as visited
            grid[i][j] = -1
            # Add the current cell to the queue for BFS
            self.queue.append((i, j))
            # Visit all adjacent cells
            list(map(dfs, (i-1, i+1, i, i), (j, j, j-1, j+1)))

        # BFS function to find the shortest bridge to the second island
        def bfs():
            dist = 0
            while self.queue:
                for _ in range(len(self.queue)):
                    i, j = self.queue.popleft()
                    for x, y in ((i-1, j), (i+1, j), (i, j-1), (i, j+1)):
                        if 0 <= x < len(grid) and 0 <= y < len(grid[0]):
                            # If we've reached the second island, return the current distance
                            if grid[x][y] == 1:
                                return dist
                            # If the current cell is water, mark it as visited and add it to the queue
                            elif not grid[x][y]:
                                grid[x][y] = -1
                                self.queue.append((x, y))
                # Increase the distance for each level we move away from the first island
                dist += 1

        self.queue = deque()
        break_flag = False
        # Find and mark the first island
        for i in range(len(grid)):
            if break_flag:
                break
            for j in range(len(grid[0])):
                if grid[i][j]:
                    dfs(i, j)
                    break_flag = True
                    break
        # Find the shortest bridge to the second island
        return bfs()


# Test case:
if __name__ == '__main__':
    Instant = Solution()
    Solve = Instant.shortestBridge(grid = [[0,1,0],[0,0,0],[0,0,1]])
    print(Solve)

File: LeetCode/Java Solutions/Shortest Bridge/shortest.java
import java.util.*;

public class Solution {
    private int[][] dirs = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
    private Queue<int[]> queue = new LinkedList<>();

    public int shortestBridge(int[][] grid) {
        int n = grid.length;

        // Start DFS to find the first island and color it with 2.
        boolean found = false;
        for (int i = 0; i < n; i++) {
            if (found) break;
            for (int j = 0; j < n; j++) {
                if (grid[i][j] == 1) {
                    dfs(grid, i, j);
                    found = true;
                    break;
                }
            }
        }

        // BFS to expand the first island until it reach the second one.
        int level = 0;
        while (!queue.isEmpty()) {
            int size = queue.size();
            while (size-- > 0) {
                int[] cur = queue.poll();
                for (int[] dir : dirs) {
                    int nx = cur[0] + dir[0];
                    int ny = cur[1] + dir[1];
                    if (nx >= 0 && ny >= 0 && nx < n && ny < n) {
                        if (grid[nx][ny] == 1) {
                            return level;
                        }
                        if (grid[nx][ny] == 0) {
                            grid[nx][ny] = 2;
                            queue.offer(new int[]{nx, ny});
                        }
                    }
                }
            }
            level++;
        }
        return -1;
    }

    private void dfs(int[][] grid, int x, int y) {
        // If out of bounds or not island, return.
        if (x < 0 || y < 0 || x >= grid.length || y >= grid.length || grid[x][y] != 1) {
            return;
        }
        // Color the island with 2 and add it into the queue.
        grid[x][y] = 2;
        queue.offer(new int[]{x, y});
        for (int[] dir : dirs) {
            dfs(grid, x + dir[0], y + dir[1]);
        }
    }

    public static void main(String[] args) {
        Solution solution = new Solution();
        int[][] grid = {{0, 1, 0}, {0, 0, 0}, {0, 0, 1}};
        System.out.println(solution.shortestBridge(grid));
    }
}



